tls/s2n_kex.c

```diff
@@ -0,0 +1,165 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "tls/s2n_server_key_exchange.h"
+#include "tls/s2n_client_key_exchange.h"
+#include "tls/s2n_kex.h"
+#include "tls/s2n_cipher_suites.h"
+#include "utils/s2n_safety.h"
+#include "s2n_tls.h"
+
+static int get_server_ecc_extension_size(const struct s2n_connection *conn)
+{
+    if (s2n_server_can_send_ec_point_formats(conn)){
+        return 6;
+    } else {
+        return 0;
+    }
+}
+
+static int get_no_extension_size(const struct s2n_connection *conn)
+{
+    return 0;
+}
+
+/* Write the Supported Points Format extension.
+ * RFC 4492 section 5.2 states that the absence of this extension in the Server Hello
+ * is equivalent to allowing only the uncompressed point format. Let's send the
+ * extension in case clients(Openssl 1.0.0) don't honor the implied behavior.
+ */
+static int write_server_ecc_extension(const struct s2n_connection *conn, struct s2n_stuffer *out)
+{
+    if (s2n_server_can_send_ec_point_formats(conn)) {
+        GUARD(s2n_stuffer_write_uint16(out, TLS_EXTENSION_EC_POINT_FORMATS));
+        /* Total extension length */
+        GUARD(s2n_stuffer_write_uint16(out, 2));
+        /* Format list length */
+        GUARD(s2n_stuffer_write_uint8(out, 1));
+        /* Only uncompressed format is supported. Interoperability shouldn't be an issue:
+         * RFC 4492 Section 5.1.2: Implementations must support it for all of their curves.
+         */
+        GUARD(s2n_stuffer_write_uint8(out, TLS_EC_FORMAT_UNCOMPRESSED));
+    }
+    return 0;
+}
+
+static int write_no_extension(const struct s2n_connection *conn, struct s2n_stuffer *out)
+{
+    return 0;
+}
+
+static int check_rsa_key(const struct s2n_connection *conn)
+{
+    return conn->config->cert_and_key_pairs != NULL;
+}
+
+static int check_dhe(const struct s2n_connection *conn)
+{
+    return conn->config->dhparams != NULL;
+}
+
+static int check_ecdhe(const struct s2n_connection *conn)
+{
+    return conn->secure.server_ecc_params.negotiated_curve != NULL;
+}
+
+const struct s2n_kex s2n_rsa = {
+        .is_ephemeral = 0,
+        .get_server_extension_size = &get_no_extension_size,
+        .write_server_extensions = &write_no_extension,
+        .connection_supported = &check_rsa_key,
+        .server_key_recv_read_data = &s2n_rsa_server_key_recv_read_data,
+        .server_key_recv_parse_data = &s2n_rsa_server_key_recv_parse_data,
+        .server_key_send = &s2n_rsa_server_key_send,
```

Comment:
Why not just make some of these RSA function pointers NULL?

Reply:
I could do that, either would end up in an error and neither should ever happen. I guess I liked seeing all the config set to real values. However this might be even more confusing and make people think they need to send an additional RSA key or call this method to do something. I'll take them out.